### PR TITLE
added conditional probability of n purchases for individual

### PR DESCRIPTION
**Pull request by MichaelSchreier**

### CODE REVIEW

File: lifetimes/fitters/pareto_nbd_fitter.py [hunk 1]
```diff
@@ -282,3 +282,64 @@ def expected_number_of_purchases_up_to_time(self, t):
         first_term = r * beta / alpha / (s - 1)
         second_term = 1 - (beta / (beta + t)) ** (s - 1)
         return first_term * second_term
+    
+        def conditional_probability_of_n_purchases_up_to_time(self, n, t, frequency, recency, T):
```
Comment (CamDavidsonPilon) [inline lifetimes/fitters/pareto_nbd_fitter.py]: @MichaelSchreier, there is a syntax error here (bad indent)

File: lifetimes/fitters/pareto_nbd_fitter.py [hunk 2]
```diff
@@ -231,9 +231,9 @@ def conditional_probability_alive(self, frequency, recency, T):
         x, t_x = frequency, recency
         r, alpha, s, beta = self._unload_params('r', 'alpha', 's', 'beta')
 
-        A_0 = np.exp(self._log_A_0([r, alpha, s, beta], x, t_x, T))
-        return 1. / (1. + (s / (r + s + x)) *
-                     (alpha + T) ** (r + x) * (beta + T) ** s * A_0)
+        A_0 = self._log_A_0([r, alpha, s, beta], x, t_x, T)
+        return 1. / (1. + exp(log(s) - log(r + s + x) +
```
Comment (CamDavidsonPilon) [inline lifetimes/fitters/pareto_nbd_fitter.py]: 👍 nice improvement

File: lifetimes/fitters/pareto_nbd_fitter.py [hunk 3]
```diff
@@ -282,3 +282,64 @@ def expected_number_of_purchases_up_to_time(self, t):
...
+        def log_B_three(i):
+            return r * log(alpha) + s * log(beta) + gammaln(r + s + x + i) + betaln(r + x + n, s + 1) + \
+                log(hyp2f1(r + s + x + i, p, r + s + x + n + 1, abs_alpha_beta/(max_of_alpha_beta + T + t))) - (
+                    gammaln(r) + gammaln(s) + (r + s + x + i) * log(max_of_alpha_beta + T + t)
+                )
+        
+        zero = (n == 0) * (1 - exp(log_p_zero))
+        one = n * log(t) - log(np.math.factorial(n)) + log_B_one - log_l
+        two = log_B_two - log_l
+        three = misc.logsumexp([i * log(t) - log(np.math.factorial(i)) + log_B_three(i) - log_l for i in range(n+1)], axis=0)
```
Comment (CamDavidsonPilon) [inline lifetimes/fitters/pareto_nbd_fitter.py]: can you pick different names here, like A, B, C, D? `one`, `two`, etc. carry too much addition meaning

**Comment by CamDavidsonPilon** [comment]
Can you please add a test to `test_estimation.py`, ideally comparing against the paper's results or another implementation?

**Comment by MichaelSchreier** [comment]
Jeez, both me and my git somehow screwed up here committing the files -.- Sorry for that!

Anyway, I fixed the indentation error (sloppy copy & paste job) and incorporated your other requests.
There are unfortunately no test cases in the source article so I went with testing the general properties that a probability should have. While writing the tests I also encountered some numerical stability issues which I then tried to fix as well.

The integration tests somehow fail on BetaGeoFitter which I did not touch here.

**Comment by CamDavidsonPilon** [comment]
Closing this in favour of #155

**Pull request closed without merging**